// amexio-chart-d3/esm5/lib/amexio-chart-d3.module.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { NgModule } from '@angular/core';
import { D3_COMPONENTS } from './amexio-chart-d3-components';
import { FormsModule } from '@angular/forms';
import { CommonModule } from '@angular/common';
import { DeviceQueryService } from './services/device.query.service';
export { DeviceQueryService } from './services/device.query.service';
var AmexioChartD3Module = /** @class */ (function () {
    function AmexioChartD3Module() {
    }
    AmexioChartD3Module.decorators = [
        { type: NgModule, args: [{
                    imports: [
                        FormsModule,
                        CommonModule
                    ],
                    declarations: D3_COMPONENTS,
                    providers: [DeviceQueryService],
                    exports: D3_COMPONENTS
                },] },
    ];
    return AmexioChartD3Module;
}());
export { AmexioChartD3Module };
