// amexio-chart-d3/esm5/lib/legend/legend.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, Input, Output, EventEmitter } from "@angular/core";
var AmexioD3Legend = /** @class */ (function () {
    function AmexioD3Legend() {
        this.horizontal = false;
        this.temparr = [];
        this.onClick = new EventEmitter();
    }
    /**
     * @return {?}
     */
    AmexioD3Legend.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
        this.data = this.data;
        this.temparr = this.data;
    };
    /**
     * @return {?}
     */
    AmexioD3Legend.prototype.ngOnChanges = /**
     * @return {?}
     */
    function () {
        // debugger;
        // window.location.reload();
        this.data = this.data;
        this.temparr = this.data;
    };
    /**
     * @param {?} node
     * @return {?}
     */
    AmexioD3Legend.prototype.onLegendClick = /**
     * @param {?} node
     * @return {?}
     */
    function (node) {
        this.onClick.emit(node);
    };
    AmexioD3Legend.decorators = [
        { type: Component, args: [{
                    selector: 'amexio-d3-legend',
                    template: "<div style =\" width: 100% ; padding-left: 20px; padding-top: 8px; \">\n<table *ngIf=\"!horizontal\" style=\"height: 100%;\">\n    <tr *ngFor = \"let node of data\" style=\"cursor: pointer;\" (click)=\"onLegendClick(node)\">\n        <td style=\"height: 10px; width: 23px; padding-left: 2px;\"\n            [ngStyle]=\"{'background-color': node.color}\"></td>\n        <td>{{node.label}}</td>\n    </tr>\n</table>\n</div>\n\n<table *ngIf=\"horizontal\" style=\"height: 100%;\">\n    <tr>\n        <td *ngFor = \"let node of data\" \n                    (click)=\"onLegendClick(node)\" \n                    style=\"cursor: pointer;\">\n            \n            <div style=\" display: inline-block; display: inline-block; padding: 2px;\">\n                    <div style=\" width: 10px; display: inline-block; padding: 2px; \" \n                    [ngStyle]=\"{'background-color': node.color}\"></div>\n                {{node.label}}\n            </div>\n        </td>\n        \n    </tr>\n</table>\n\n<br>\n "
                },] },
    ];
    AmexioD3Legend.ctorParameters = function () { return []; };
    AmexioD3Legend.propDecorators = {
        data: [{ type: Input, args: ['data',] }],
        horizontal: [{ type: Input, args: ['horizontal',] }],
        onClick: [{ type: Output }]
    };
    return AmexioD3Legend;
}());
export { AmexioD3Legend };
if (false) {
    /** @type {?} */
    AmexioD3Legend.prototype.data;
    /** @type {?} */
    AmexioD3Legend.prototype.horizontal;
    /** @type {?} */
    AmexioD3Legend.prototype.temparr;
    /** @type {?} */
    AmexioD3Legend.prototype.onClick;
}
