// amexio-chart-d3/esm2015/lib/pie/pie.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, Input, Output, EventEmitter, ViewChild } from '@angular/core';
export class AmexioD3PieChartComponent {
    constructor() {
        this.data = [];
        this.onLegendClick = new EventEmitter();
        this.onChartClick = new EventEmitter();
        this.drillableEvent = new EventEmitter();
        this.legend = true;
        this.colors = [];
        this.width = "300";
        this.height = "300";
        this.title = "";
        this.level = 0;
        this.labelcolor = "black";
        this.labelflag = false;
        this.drillableFlag = true;
    }
    /**
     * @return {?}
     */
    ngOnInit() {
    }
    /**
     * @param {?} node
     * @return {?}
     */
    legendClick(node) {
        this.onLegendClick.emit(node);
    }
    /**
     * @param {?} node
     * @return {?}
     */
    chartClick(node) {
        this.onChartClick.emit(node);
    }
    /**
     * @param {?} ref
     * @return {?}
     */
    fordrillableClick(ref) {
        this.drillableEvent.emit(ref);
    }
    /**
     * @param {?} data
     * @return {?}
     */
    fetchData(data) {
        this.drillData = data;
    }
    /**
     * @return {?}
     */
    drawChart() {
        this.donutref.drillableFlag = true;
        this.donutref.drawChart();
    }
}
AmexioD3PieChartComponent.decorators = [
    { type: Component, args: [{
                selector: 'amexio-d3-chart-pie',
                template: `<div   *ngIf="drillableFlag">
<amexio-d3-chart-donut #id (onLegendClick)="legendClick($event)" 
                       (onChartClick)="chartClick($event)" 
                       (drillableEvent)=" fordrillableClick($event)"
                        [height]="height"
                        [width]="width"
                        [title]="title"
                        [data] = "data"
                        [pie]  = "true"
                        [color]="colors"
                        [http-url]="httpurl"
                        [http-method]="httpmethod"
                        [data-reader]="datareader"
                        [label]="labelflag"
                        [label-color]="labelcolor"
                        [level]="level"
                        [target]="target"
                        [drillable-data]="drillabledatakey"
                        [drillData]="drillData">
</amexio-d3-chart-donut>
</div>
`
            },] },
];
AmexioD3PieChartComponent.propDecorators = {
    data: [{ type: Input, args: ['data',] }],
    onLegendClick: [{ type: Output }],
    onChartClick: [{ type: Output }],
    drillableEvent: [{ type: Output }],
    donutref: [{ type: ViewChild, args: ['id',] }],
    legend: [{ type: Input }],
    colors: [{ type: Input, args: ['color',] }],
    width: [{ type: Input, args: ['width',] }],
    height: [{ type: Input, args: ['height',] }],
    title: [{ type: Input }],
    httpurl: [{ type: Input, args: ['http-url',] }],
    httpmethod: [{ type: Input, args: ['http-method',] }],
    datareader: [{ type: Input, args: ['data-reader',] }],
    level: [{ type: Input, args: ['level',] }],
    target: [{ type: Input, args: ['target',] }],
    drillabledatakey: [{ type: Input, args: ['drillable-data',] }],
    labelcolor: [{ type: Input, args: ['label-color',] }],
    labelflag: [{ type: Input, args: ['label',] }]
};
if (false) {
    /** @type {?} */
    AmexioD3PieChartComponent.prototype.data;
    /** @type {?} */
    AmexioD3PieChartComponent.prototype.onLegendClick;
    /** @type {?} */
    AmexioD3PieChartComponent.prototype.onChartClick;
    /** @type {?} */
    AmexioD3PieChartComponent.prototype.drillableEvent;
    /** @type {?} */
    AmexioD3PieChartComponent.prototype.donutref;
    /** @type {?} */
    AmexioD3PieChartComponent.prototype.legend;
    /** @type {?} */
    AmexioD3PieChartComponent.prototype.colors;
    /** @type {?} */
    AmexioD3PieChartComponent.prototype.width;
    /** @type {?} */
    AmexioD3PieChartComponent.prototype.height;
    /** @type {?} */
    AmexioD3PieChartComponent.prototype.title;
    /** @type {?} */
    AmexioD3PieChartComponent.prototype.httpurl;
    /** @type {?} */
    AmexioD3PieChartComponent.prototype.httpmethod;
    /** @type {?} */
    AmexioD3PieChartComponent.prototype.datareader;
    /** @type {?} */
    AmexioD3PieChartComponent.prototype.level;
    /** @type {?} */
    AmexioD3PieChartComponent.prototype.target;
    /** @type {?} */
    AmexioD3PieChartComponent.prototype.drillabledatakey;
    /** @type {?} */
    AmexioD3PieChartComponent.prototype.labelcolor;
    /** @type {?} */
    AmexioD3PieChartComponent.prototype.labelflag;
    /** @type {?} */
    AmexioD3PieChartComponent.prototype.drillableFlag;
    /** @type {?} */
    AmexioD3PieChartComponent.prototype.drillData;
}
