// amexio-chart-d3/esm2015/lib/drillable/drillable/drillable.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component } from '@angular/core';
import { ViewDrillableComponent } from '../../base/view.component';
export class DrillableComponent extends ViewDrillableComponent {
    constructor() {
        super();
        this.mouseLocation = { left: 0, top: 0 };
        this.componentcounter = 1;
        this.disableFlag = false;
    }
    /**
     * @return {?}
     */
    ngOnInit() {
    }
    /**
     * @return {?}
     */
    ngAfterViewInit() {
        this.ComponentDataArray = this.getComponentData();
        this.getData(this.ComponentDataArray);
    }
    /**
     * @param {?} data
     * @return {?}
     */
    getData(data) {
        data.forEach((/**
         * @param {?} element
         * @return {?}
         */
        element => {
            if (element.level !== 1) {
                element.drillableFlag = false;
                element.drillableEvent.subscribe((/**
                 * @param {?} eventdata
                 * @return {?}
                 */
                (eventdata) => {
                    this.hideUI(eventdata);
                    this.sameLevelChart(eventdata);
                    this.ShowUI(eventdata);
                }));
            }
            else {
                this.disableFlag = false;
                element.drillableEvent.subscribe((/**
                 * @param {?} eventdata
                 * @return {?}
                 */
                (eventdata) => {
                    this.hideUI(eventdata);
                    this.sameLevelChart(eventdata);
                    this.ShowUI(eventdata);
                }));
            }
        }));
    }
    /**
     * @param {?} eventRef
     * @return {?}
     */
    hideUI(eventRef) {
        this.ComponentDataArray.forEach((/**
         * @param {?} element
         * @return {?}
         */
        element => {
            if (element.level == eventRef.ref.level) {
                eventRef.ref.drillableFlag = false;
            }
        }));
    }
    /**
     * @param {?} eventRef
     * @return {?}
     */
    ShowUI(eventRef) {
        this.eventData = [];
        this.eventData = eventRef;
        this.ComponentDataArray.forEach((/**
         * @param {?} element
         * @return {?}
         */
        element => {
            if (eventRef.ref.target == element.level) {
                if (this.arrayofsameLevel.length == 1) {
                    this.disableFlag = true;
                    element.drillableFlag = true;
                    element.fetchData(eventRef.node);
                    this.componentcounter = element.level;
                }
                else {
                    this.contextmenuFlag = true;
                    this.mouseLocation.left = eventRef.event.clientX;
                    this.mouseLocation.top = eventRef.event.clientY;
                    this.contextStyle = this.getContextMenuStyle();
                }
            }
        }));
        if (this.componentcounter == eventRef.ref.level) {
            eventRef.ref.drillableFlag = true;
        }
    }
    /**
     * @param {?} event
     * @return {?}
     */
    previouschartClick(event) {
        /** @type {?} */
        let chartLevel = this.componentcounter - 1;
        this.ComponentDataArray.forEach((/**
         * @param {?} element
         * @return {?}
         */
        element => {
            if (chartLevel == element.level) {
                element.drillableFlag = true;
                element.drawChart();
            }
            else if (this.componentcounter == element.level) {
                element.drillableFlag = false;
            }
        }));
        if (this.componentcounter == 2) {
            this.disableFlag = false;
        }
        this.componentcounter = chartLevel;
    }
    /**
     * @param {?} eventRef
     * @return {?}
     */
    sameLevelChart(eventRef) {
        this.arrayofsameLevel = [];
        this.ComponentDataArray.forEach((/**
         * @param {?} element
         * @return {?}
         */
        element => {
            if (eventRef.ref.target == element.level) {
                this.arrayofsameLevel = this.arrayofsameLevel.concat(element);
            }
        }));
    }
    //Method Creates style if two same targets found.
    /**
     * @return {?}
     */
    getContextMenuStyle() {
        return {
            'cursor': 'default',
            'position': 'fixed',
            'display': this.contextmenuFlag ? 'block' : 'none',
            'left': this.mouseLocation.left + 'px',
            'top': this.mouseLocation.top + 'px',
            'box-shadow': '1px 1px 2px #000000',
            'width': '15%',
        };
    }
    /**
     * @param {?} elementRef
     * @return {?}
     */
    getListPosition(elementRef) {
        /** @type {?} */
        const height = 240;
        if ((window.screen.height - elementRef.getBoundingClientRect().bottom) < height) {
            return true;
        }
        else {
            return false;
        }
    }
    /**
     * @param {?} event
     * @return {?}
     */
    onContextNodeClick(event) {
        this.componentcounter = 0;
        this.contextmenuFlag = false;
        this.eventData.ref.drillableFlag = false;
        this.disableFlag = true;
        event.drillableFlag = true;
        event.fetchData(this.eventData.node);
        this.componentcounter = event.level;
        if (this.componentcounter == this.eventData.ref.level) {
            this.eventData.ref.drillableFlag = true;
        }
    }
    /**
     * @return {?}
     */
    onCloseClick() {
        this.contextmenuFlag = false;
    }
}
DrillableComponent.decorators = [
    { type: Component, args: [{
                selector: 'amexio-drillable',
                template: `
<div > 
    <ng-container *ngIf="disableFlag">
            <button type="button" class="button"
             (click)="previouschartClick($event)">
             <i class="fa fa-arrow-left"></i> 
             
            </button>
</ng-container>

<ng-content></ng-content>

</div>

<span [ngStyle]="contextStyle" >
  <ul *ngIf="contextmenuFlag" class="close-thik" class="context-menu-list"  >
    <span (click)="onCloseClick()" class="close"><i class="fa fa-close"></i></span>
  <li (click)="onContextNodeClick(itemConfig)" class="context-menu-list-items" 
  *ngFor="let itemConfig of arrayofsameLevel">
  <em [ngStyle]="{'padding-left': itemConfig.icon ? '5px':'19px'}"></em>
  <span style="white-space: nowrap;display: inline ; padding-left:5px">{{itemConfig.title}}
  </span>
  </li>
  </ul>
  </span>
  `,
                styles: [`.button{position:relative;border:1px solid #555;font-weight:400;color:#000;padding:.5rem 1rem;text-align:center;transition-duration:.4s;text-decoration:none;overflow:hidden;cursor:pointer;border-radius:3px}.button:after{content:"";background:#000;display:block;position:absolute;padding-top:300%;padding-left:350%;margin-left:-20px!important;margin-top:-120%;opacity:0;transition:.8s}.button:active:after{padding:0;margin:0;opacity:1;transition:none}.context-menu-list{display:block;position:absolute;flex-direction:column;min-width:180px;cursor:pointer;max-height:150px;z-index:1;padding:10px 0;overflow-y:auto;background-color:#fff!important;box-shadow:0 5px 5px -3px rgba(0,0,0,.2),0 8px 10px 1px rgba(0,0,0,.14),0 3px 14px 2px rgba(0,0,0,.12)}.context-menu-list .context-menu-list-items{text-decoration:none;font-size:14px!important;padding:5px;display:block}.close{float:right;padding-right:8px;cursor:pointer}`]
            },] },
];
DrillableComponent.ctorParameters = () => [];
if (false) {
    /** @type {?} */
    DrillableComponent.prototype.mouseLocation;
    /** @type {?} */
    DrillableComponent.prototype.contextmenuFlag;
    /** @type {?} */
    DrillableComponent.prototype.contextStyle;
    /** @type {?} */
    DrillableComponent.prototype.ComponentDataArray;
    /** @type {?} */
    DrillableComponent.prototype.componentcounter;
    /** @type {?} */
    DrillableComponent.prototype.disableFlag;
    /** @type {?} */
    DrillableComponent.prototype.arrayofsameLevel;
    /** @type {?} */
    DrillableComponent.prototype.eventData;
    /** @type {?} */
    DrillableComponent.prototype.rightClickNodeData;
    /** @type {?} */
    DrillableComponent.prototype.posixUp;
    /** @type {?} */
    DrillableComponent.prototype.levelArray;
}
