// amexio-chart-d3/esm5/lib/pie/pie.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, Input, Output, EventEmitter, ViewChild } from '@angular/core';
var AmexioD3PieChartComponent = /** @class */ (function () {
    function AmexioD3PieChartComponent() {
        this.data = [];
        this.onLegendClick = new EventEmitter();
        this.onChartClick = new EventEmitter();
        this.drillableEvent = new EventEmitter();
        this.legend = true;
        this.colors = [];
        this.width = "300";
        this.height = "300";
        this.title = "";
        this.level = 0;
        this.labelcolor = "black";
        this.labelflag = false;
        this.drillableFlag = true;
    }
    /**
     * @return {?}
     */
    AmexioD3PieChartComponent.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
    };
    /**
     * @param {?} node
     * @return {?}
     */
    AmexioD3PieChartComponent.prototype.legendClick = /**
     * @param {?} node
     * @return {?}
     */
    function (node) {
        this.onLegendClick.emit(node);
    };
    /**
     * @param {?} node
     * @return {?}
     */
    AmexioD3PieChartComponent.prototype.chartClick = /**
     * @param {?} node
     * @return {?}
     */
    function (node) {
        this.onChartClick.emit(node);
    };
    /**
     * @param {?} ref
     * @return {?}
     */
    AmexioD3PieChartComponent.prototype.fordrillableClick = /**
     * @param {?} ref
     * @return {?}
     */
    function (ref) {
        this.drillableEvent.emit(ref);
    };
    /**
     * @param {?} data
     * @return {?}
     */
    AmexioD3PieChartComponent.prototype.fetchData = /**
     * @param {?} data
     * @return {?}
     */
    function (data) {
        this.drillData = data;
    };
    /**
     * @return {?}
     */
    AmexioD3PieChartComponent.prototype.drawChart = /**
     * @return {?}
     */
    function () {
        this.donutref.drillableFlag = true;
        this.donutref.drawChart();
    };
    AmexioD3PieChartComponent.decorators = [
        { type: Component, args: [{
                    selector: 'amexio-d3-chart-pie',
                    template: "<div   *ngIf=\"drillableFlag\">\n<amexio-d3-chart-donut #id (onLegendClick)=\"legendClick($event)\" \n                       (onChartClick)=\"chartClick($event)\" \n                       (drillableEvent)=\" fordrillableClick($event)\"\n                        [height]=\"height\"\n                        [width]=\"width\"\n                        [title]=\"title\"\n                        [data] = \"data\"\n                        [pie]  = \"true\"\n                        [color]=\"colors\"\n                        [http-url]=\"httpurl\"\n                        [http-method]=\"httpmethod\"\n                        [data-reader]=\"datareader\"\n                        [label]=\"labelflag\"\n                        [label-color]=\"labelcolor\"\n                        [level]=\"level\"\n                        [target]=\"target\"\n                        [drillable-data]=\"drillabledatakey\"\n                        [drillData]=\"drillData\">\n</amexio-d3-chart-donut>\n</div>\n"
                },] },
    ];
    AmexioD3PieChartComponent.propDecorators = {
        data: [{ type: Input, args: ['data',] }],
        onLegendClick: [{ type: Output }],
        onChartClick: [{ type: Output }],
        drillableEvent: [{ type: Output }],
        donutref: [{ type: ViewChild, args: ['id',] }],
        legend: [{ type: Input }],
        colors: [{ type: Input, args: ['color',] }],
        width: [{ type: Input, args: ['width',] }],
        height: [{ type: Input, args: ['height',] }],
        title: [{ type: Input }],
        httpurl: [{ type: Input, args: ['http-url',] }],
        httpmethod: [{ type: Input, args: ['http-method',] }],
        datareader: [{ type: Input, args: ['data-reader',] }],
        level: [{ type: Input, args: ['level',] }],
        target: [{ type: Input, args: ['target',] }],
        drillabledatakey: [{ type: Input, args: ['drillable-data',] }],
        labelcolor: [{ type: Input, args: ['label-color',] }],
        labelflag: [{ type: Input, args: ['label',] }]
    };
    return AmexioD3PieChartComponent;
}());
export { AmexioD3PieChartComponent };
if (false) {
    /** @type {?} */
    AmexioD3PieChartComponent.prototype.data;
    /** @type {?} */
    AmexioD3PieChartComponent.prototype.onLegendClick;
    /** @type {?} */
    AmexioD3PieChartComponent.prototype.onChartClick;
    /** @type {?} */
    AmexioD3PieChartComponent.prototype.drillableEvent;
    /** @type {?} */
    AmexioD3PieChartComponent.prototype.donutref;
    /** @type {?} */
    AmexioD3PieChartComponent.prototype.legend;
    /** @type {?} */
    AmexioD3PieChartComponent.prototype.colors;
    /** @type {?} */
    AmexioD3PieChartComponent.prototype.width;
    /** @type {?} */
    AmexioD3PieChartComponent.prototype.height;
    /** @type {?} */
    AmexioD3PieChartComponent.prototype.title;
    /** @type {?} */
    AmexioD3PieChartComponent.prototype.httpurl;
    /** @type {?} */
    AmexioD3PieChartComponent.prototype.httpmethod;
    /** @type {?} */
    AmexioD3PieChartComponent.prototype.datareader;
    /** @type {?} */
    AmexioD3PieChartComponent.prototype.level;
    /** @type {?} */
    AmexioD3PieChartComponent.prototype.target;
    /** @type {?} */
    AmexioD3PieChartComponent.prototype.drillabledatakey;
    /** @type {?} */
    AmexioD3PieChartComponent.prototype.labelcolor;
    /** @type {?} */
    AmexioD3PieChartComponent.prototype.labelflag;
    /** @type {?} */
    AmexioD3PieChartComponent.prototype.drillableFlag;
    /** @type {?} */
    AmexioD3PieChartComponent.prototype.drillData;
}
