// amexio-chart-d3/esm5/lib/candlestick/candlestick.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { Component, Input, ViewChild, ElementRef, Output, EventEmitter, ChangeDetectorRef, } from "@angular/core";
import { AmexioD3BaseChartComponent } from "../base/base.component";
import { CommanDataService } from '../services/comman.data.service';
import { DeviceQueryService } from '../services/device.query.service';
import * as d3 from 'd3';
var CandlestickComponent = /** @class */ (function (_super) {
    tslib_1.__extends(CandlestickComponent, _super);
    function CandlestickComponent(myservice, cdf, device) {
        var _this = _super.call(this, "candlestickwaterfallchart") || this;
        _this.myservice = myservice;
        _this.cdf = cdf;
        _this.device = device;
        _this.svgheight = 300;
        _this.showzeroflag = true;
        _this.onLegendClick = new EventEmitter();
        _this.predefinedColor = [];
        _this.keyArray = [];
        _this.margin = {};
        _this.legendArray = [];
        return _this;
    }
    /**
     * @return {?}
     */
    CandlestickComponent.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
        var _this = this;
        this.wt = this.svgwidth;
        this.predefinedColor = ["#3366cc", "#dc3912", "#ff9900", "#109618", "#990099", "#0099c6", "#dd4477", "#66aa00", "#b82e2e", "#316395", "#994499", "#22aa99", "#aaaa11", "#6633cc", "#e67300", "#8b0707", "#651067", "#329262", "#5574a6", "#3b3eac"];
        if (this.level <= 1) {
            /** @type {?} */
            var res = void 0;
            if (this.httpmethod && this.httpurl) {
                this.myservice.fetchUrlData(this.httpurl, this.httpmethod).subscribe((/**
                 * @param {?} response
                 * @return {?}
                 */
                function (response) {
                    //this.data = response;
                    _this.httpresponse = response;
                    _this.data = _this.getResponseData(response);
                }), (/**
                 * @param {?} error
                 * @return {?}
                 */
                function (error) {
                }), (/**
                 * @return {?}
                 */
                function () {
                    setTimeout((/**
                     * @return {?}
                     */
                    function () {
                        _this.transformData(_this.data);
                        _this.initializeData();
                        _this.plotXYAxis();
                        _this.plotD3Chart();
                    }), 0);
                }));
            }
            else if (this.data) {
                setTimeout((/**
                 * @return {?}
                 */
                function () {
                    _this.transformData(_this.data);
                    _this.initializeData();
                    _this.plotXYAxis();
                    _this.plotD3Chart();
                }), 0);
            }
        }
    };
    /**
     * @param {?} data
     * @return {?}
     */
    CandlestickComponent.prototype.fetchData = /**
     * @param {?} data
     * @return {?}
     */
    function (data) {
        var _this = this;
        /** @type {?} */
        var requestJson;
        /** @type {?} */
        var key = this.drillabledatakey;
        /** @type {?} */
        var resp;
        if (this.drillabledatakey.length) {
            /** @type {?} */
            var drillabledata = this.getMultipleDrillbleKeyData(data, key);
            requestJson = drillabledata;
        }
        else {
            requestJson = data;
        }
        if (this.httpmethod && this.httpurl) {
            this.myservice.postfetchData(this.httpurl, this.httpmethod, requestJson).subscribe((/**
             * @param {?} response
             * @return {?}
             */
            function (response) {
                resp = response;
                _this.httpresponse = response;
            }), (/**
             * @param {?} error
             * @return {?}
             */
            function (error) {
            }), (/**
             * @return {?}
             */
            function () {
                setTimeout((/**
                 * @return {?}
                 */
                function () {
                    //this.data = this.getResponseData(resp);
                    _this.drawChart();
                }), 0);
            }));
        }
    };
    /**
     * @return {?}
     */
    CandlestickComponent.prototype.drawChart = /**
     * @return {?}
     */
    function () {
        var _this = this;
        setTimeout((/**
         * @return {?}
         */
        function () {
            _this.data = _this.getResponseData(_this.httpresponse);
            _this.transformData(_this.data);
            _this.initializeData();
            _this.plotXYAxis();
            _this.plotD3Chart();
        }), 0);
    };
    /**
     * @return {?}
     */
    CandlestickComponent.prototype.initializeData = /**
     * @return {?}
     */
    function () {
        this.tooltip = this.toolTip(d3);
        if (this.resizeflag == false) {
            //RESIZE STEP 1
            if (this.wt) {
                this.svgwidth = this.wt;
            }
            else if (this.chartId) {
                this.svgwidth = this.chartId.nativeElement.offsetWidth;
            }
            //RESIZE STEP 1 ENDS HERE 
        }
        this.margin = { top: 20, right: 30, bottom: 50, left: 60 },
            this.width = this.svgwidth - this.margin.left - this.margin.right,
            this.height = this.svgheight - this.margin.top - this.margin.bottom;
    };
    /**
     * @return {?}
     */
    CandlestickComponent.prototype.plotXYAxis = /**
     * @return {?}
     */
    function () {
        // set the ranges
        this.x = d3.scaleBand().range([0, this.width]);
        this.y = d3.scaleLinear()
            .rangeRound([this.height, 0]);
        // scale the range of the data
        /** @type {?} */
        var candlestickArray = this.data.map((/**
         * @param {?} d
         * @return {?}
         */
        function (d) {
            return d[Object.keys(d)[0]];
        }));
        this.x.domain(candlestickArray);
        /** @type {?} */
        var max = d3.max(this.data, (/**
         * @param {?} d
         * @return {?}
         */
        function (d) { return d.end; }));
        this.y.domain([0, max]);
        // this.svg = d3.select("#" + this.componentId)
        //RESIZE STEP 2 START
        this.svg = d3.select("#" + this.componentId)
            .attr('viewBox', '0 0 ' + this.svgwidth + ' ' + this.svgheight)
            .attr("width", this.width + this.margin.left + this.margin.right)
            .attr("height", this.height + this.margin.top + this.margin.bottom)
            .append("g")
            .attr("transform", "translate(" + this.margin.left + "," + this.margin.top + ")");
        // add the X Axis
        if (this.device.IsDesktop() == true) {
            if (this.svgwidth <= 400) {
                this.svg.append("g")
                    .attr("transform", "translate(0," + this.height + ")")
                    .call(d3.axisBottom(this.x)).
                    selectAll("text")
                    .attr("y", 0)
                    .attr("x", 9)
                    .attr("dy", ".35em")
                    .attr("transform", "rotate(60)")
                    .style("text-anchor", "start");
            }
            else {
                this.svg.append("g")
                    .attr("transform", "translate(0," + this.height + ")")
                    .call(d3.axisBottom(this.x));
            }
        }
        else {
            this.svg.append("g")
                .attr("transform", "translate(0," + this.height + ")")
                .call(d3.axisBottom(this.x)).
                selectAll("text")
                .attr("y", 0)
                .attr("x", 9)
                .attr("dy", ".35em")
                .attr("transform", "rotate(60)")
                .style("text-anchor", "start");
        }
        // this.svg.append("g")
        //   .attr("transform", "translate(0," + this.height + ")")
        //   .call(d3.axisBottom(this.x));
        // add the Y Axis
        this.svg.append("g")
            .call(d3.axisLeft(this.y).ticks(this.tickscount));
        this.plotLine(this.svg, this.x, this.y, this.height, this.width);
    };
    /**
     * @return {?}
     */
    CandlestickComponent.prototype.plotD3Chart = /**
     * @return {?}
     */
    function () {
        var _this = this;
        /** @type {?} */
        var bar = this.svg.selectAll(".bar")
            .data(this.data)
            .enter().append("g")
            .attr("class", (/**
         * @param {?} d
         * @return {?}
         */
        function (d) { return "bar " + d.class; }))
            .attr("transform", (/**
         * @param {?} d
         * @return {?}
         */
        function (d) {
            return "translate(" + _this.x(d[Object.keys(d)[0]]) + ",0)";
        }));
        bar.append("rect")
            .attr("y", (/**
         * @param {?} d
         * @return {?}
         */
        function (d) { return _this.y(Math.max(d.start, d.end)); }))
            .attr("height", (/**
         * @param {?} d
         * @return {?}
         */
        function (d) { return Math.abs(_this.y(d.start) - _this.y(d.end)); }))
            .attr("width", this.x.bandwidth())
            .attr("fill", (/**
         * @param {?} d
         * @param {?} i
         * @return {?}
         */
        function (d, i) {
            return _this.predefinedColor[i];
        }))
            .attr("cursor", "pointer")
            .on("mouseover", (/**
         * @param {?} d
         * @return {?}
         */
        function (d) {
            _this.formTooltipData(d);
            return _this.tooltip.style("visibility", "visible");
        }))
            .on("mousemove", (/**
         * @param {?} d
         * @return {?}
         */
        function (d) {
            return _this.tooltip.html(_this.formTooltipData(d))
                .style("top", (d3.event.pageY - 10) + "px")
                .style("left", (d3.event.pageX + 10) + "px");
        }))
            .on("mouseout", (/**
         * @param {?} d
         * @return {?}
         */
        function (d) {
            return _this.tooltip.style("visibility", "hidden");
        }))
            .on("click", (/**
         * @param {?} d
         * @return {?}
         */
        function (d) {
            _this.onCandlestickClick(d);
            _this.fordrillableClick(_this, d, event);
            return _this.tooltip.style("visibility", "hidden");
        }));
        if (this.labelflag) {
            bar.append("text")
                .style("font-weight", "bold")
                .style("font-size", "1vw")
                .attr("text-anchor", "middle")
                .attr("fill", (/**
             * @param {?} d
             * @return {?}
             */
            function (d) {
                if (_this.labelcolor && _this.labelcolor.length > 0) {
                    return _this.labelcolor;
                }
                else {
                    return "black";
                }
            }))
                .attr("x", (/**
             * @param {?} d
             * @return {?}
             */
            function (d) {
                return (_this.x.bandwidth()) / 2;
            }))
                .attr("y", (/**
             * @param {?} d
             * @param {?} index
             * @return {?}
             */
            function (d, index) {
                return _this.y(Math.max(d.start, d.end)) + 20;
            }))
                .text((/**
             * @param {?} d
             * @return {?}
             */
            function (d) {
                // return d[Object.keys(d)[1]]
                if (_this.showzeroflag) {
                    return d[Object.keys(d)[1]];
                }
                else if (!_this.showzeroflag) {
                    if (d[Object.keys(d)[1]] > 0) {
                        return d[Object.keys(d)[1]];
                    }
                }
            }))
                .attr("cursor", "pointer")
                .on("mouseover", (/**
             * @param {?} d
             * @return {?}
             */
            function (d) {
                _this.formTooltipData(d);
                return _this.tooltip.style("visibility", "visible");
            }))
                .on("mousemove", (/**
             * @param {?} d
             * @return {?}
             */
            function (d) {
                return _this.tooltip.html(_this.formTooltipData(d))
                    .style("top", (d3.event.pageY - 10) + "px")
                    .style("left", (d3.event.pageX + 10) + "px");
            }))
                .on("mouseout", (/**
             * @param {?} d
             * @return {?}
             */
            function (d) {
                return _this.tooltip.style("visibility", "hidden");
            }))
                .on("click", (/**
             * @param {?} d
             * @return {?}
             */
            function (d) {
                _this.onCandlestickClick(d);
                _this.fordrillableClick(_this, d, event);
                return _this.tooltip.style("visibility", "hidden");
            }));
        }
    };
    /**
     * @param {?} svg
     * @param {?} x
     * @param {?} y
     * @param {?} height
     * @param {?} width
     * @return {?}
     */
    CandlestickComponent.prototype.plotLine = /**
     * @param {?} svg
     * @param {?} x
     * @param {?} y
     * @param {?} height
     * @param {?} width
     * @return {?}
     */
    function (svg, x, y, height, width) {
        if (this.hScale) {
            svg.append('g')
                .attr("color", "lightgrey")
                .call(d3.axisLeft(y)
                .tickSize(-width).tickFormat('').ticks(this.tickscount));
        }
    };
    /**
     * @param {?} data
     * @return {?}
     */
    CandlestickComponent.prototype.transformData = /**
     * @param {?} data
     * @return {?}
     */
    function (data) {
        var _this = this;
        this.transformeddata = [];
        this.keyArray = [];
        this.keyArray = data[0];
        data.forEach((/**
         * @param {?} element
         * @param {?} index
         * @return {?}
         */
        function (element, index) {
            if (index > 0) {
                /** @type {?} */
                var DummyObject_1 = {};
                element.forEach((/**
                 * @param {?} individualvalue
                 * @param {?} keyindex
                 * @return {?}
                 */
                function (individualvalue, keyindex) {
                    DummyObject_1[_this.keyArray[keyindex]] = individualvalue;
                })); //inner for loop ends
                _this.transformeddata.push(DummyObject_1);
            } //if ends
        })); //outer for loop ends 
        this.data = this.transformeddata;
        this.addDataKeys();
        this.formLegendData();
    };
    /**
     * @return {?}
     */
    CandlestickComponent.prototype.addDataKeys = /**
     * @return {?}
     */
    function () {
        /** @type {?} */
        var cumulative = 0;
        for (var i = 0; i < this.data.length; i++) {
            this.data[i]["start"] = cumulative;
            cumulative += this.data[i][this.keyArray[1]];
            this.data[i]["end"] = cumulative;
            this.data[i]["class"] = (this.data[i][this.keyArray[1]] >= 0) ? 'positive' : 'negative';
        }
    };
    /**
     * @param {?} httpResponse
     * @return {?}
     */
    CandlestickComponent.prototype.getResponseData = /**
     * @param {?} httpResponse
     * @return {?}
     */
    function (httpResponse) {
        /** @type {?} */
        var responsedata = httpResponse;
        if (this.datareader != null) {
            /** @type {?} */
            var dr = this.datareader.split('.');
            try {
                for (var dr_1 = tslib_1.__values(dr), dr_1_1 = dr_1.next(); !dr_1_1.done; dr_1_1 = dr_1.next()) {
                    var ir = dr_1_1.value;
                    responsedata = responsedata[ir];
                }
            }
            catch (e_1_1) { e_1 = { error: e_1_1 }; }
            finally {
                try {
                    if (dr_1_1 && !dr_1_1.done && (_a = dr_1.return)) _a.call(dr_1);
                }
                finally { if (e_1) throw e_1.error; }
            }
        }
        else {
            responsedata = httpResponse;
        }
        return responsedata;
        var e_1, _a;
    };
    /**
     * @param {?} tooltipData
     * @return {?}
     */
    CandlestickComponent.prototype.formTooltipData = /**
     * @param {?} tooltipData
     * @return {?}
     */
    function (tooltipData) {
        /** @type {?} */
        var object = {};
        object[this.keyArray[0]] = tooltipData[Object.keys(tooltipData)[0]];
        object[this.keyArray[1]] = tooltipData[Object.keys(tooltipData)[1]];
        return this.toolTipForBar(object);
    };
    /**
     * @param {?} chartData
     * @return {?}
     */
    CandlestickComponent.prototype.onCandlestickClick = /**
     * @param {?} chartData
     * @return {?}
     */
    function (chartData) {
        /** @type {?} */
        var object = {};
        object[this.keyArray[0]] = chartData[Object.keys(chartData)[0]];
        object[this.keyArray[1]] = chartData[Object.keys(chartData)[1]];
        this.chartClick(object);
    };
    /**
     * @return {?}
     */
    CandlestickComponent.prototype.formLegendData = /**
     * @return {?}
     */
    function () {
        var _this = this;
        this.legendArray = [];
        this.data.forEach((/**
         * @param {?} element
         * @param {?} index
         * @return {?}
         */
        function (element, index) {
            try {
                for (var _a = tslib_1.__values(Object.entries(element)), _b = _a.next(); !_b.done; _b = _a.next()) {
                    var _c = tslib_1.__read(_b.value, 2), key = _c[0], value = _c[1];
                    if (key == _this.keyArray[0]) {
                        /** @type {?} */
                        var object = {};
                        object["label"] = value;
                        object["color"] = _this.predefinedColor[index];
                        _this.legendArray.push(object);
                    }
                }
            }
            catch (e_2_1) { e_2 = { error: e_2_1 }; }
            finally {
                try {
                    if (_b && !_b.done && (_d = _a.return)) _d.call(_a);
                }
                finally { if (e_2) throw e_2.error; }
            }
            var e_2, _d;
        }));
    };
    /**
     * @param {?} chartData
     * @return {?}
     */
    CandlestickComponent.prototype.onCandlestickLegendClick = /**
     * @param {?} chartData
     * @return {?}
     */
    function (chartData) {
        /** @type {?} */
        var object = {};
        this.data.forEach((/**
         * @param {?} element
         * @return {?}
         */
        function (element) {
            try {
                for (var _a = tslib_1.__values(Object.entries(element)), _b = _a.next(); !_b.done; _b = _a.next()) {
                    var _c = tslib_1.__read(_b.value, 2), key = _c[0], value = _c[1];
                    if (value == chartData.label) {
                        object[chartData.label] = element[Object.keys(element)[1]];
                    }
                }
            }
            catch (e_3_1) { e_3 = { error: e_3_1 }; }
            finally {
                try {
                    if (_b && !_b.done && (_d = _a.return)) _d.call(_a);
                }
                finally { if (e_3) throw e_3.error; }
            }
            var e_3, _d;
        }));
        this.onLegendClick.emit(object);
    };
    //RESIZE STEP 4 STARTS
    //RESIZE STEP 4 STARTS
    /**
     * @return {?}
     */
    CandlestickComponent.prototype.validateresize = 
    //RESIZE STEP 4 STARTS
    /**
     * @return {?}
     */
    function () {
        var _this = this;
        setTimeout((/**
         * @return {?}
         */
        function () {
            if (_this.wt) {
            }
            else {
                _this.resize();
            }
        }), 0);
    };
    //RESIZE STEP 4 ENDS
    //RESIZE STEP 4 ENDS
    /**
     * @return {?}
     */
    CandlestickComponent.prototype.resize = 
    //RESIZE STEP 4 ENDS
    /**
     * @return {?}
     */
    function () {
        this.svg.selectAll("*").remove();
        this.resizeflag = true;
        if (this.wt) {
            this.svgwidth = this.wt;
        }
        else if (this.chartId) {
            // this.resizewt = this.chartId.nativeElement.offsetWidth;
            // console.log("", new Date().getTime(), " ", this.resizewt);
            this.svgwidth = this.chartId.nativeElement.offsetWidth;
        }
        this.cdf.detectChanges();
        this.initializeData();
        this.plotXYAxis();
        this.plotD3Chart();
    };
    CandlestickComponent.decorators = [
        { type: Component, args: [{
                    selector: 'amexio-d3-chart-waterfall',
                    template: "\n<div #chartId *ngIf=\"drillableFlag\"  (window:resize)=\"validateresize()\" >\n<table *ngIf=\"data.length > 0\" style=\"width: 100%;\" >\n  <tr *ngIf = \"title.length > 0\">\n      <td style=\"font-size: 18px; font-weight: bold; \"> {{title}} </td>\n  </tr>\n  <tr>\n      <td>\n          <svg [attr.height]=\"svgheight\" [attr.width]=\"svgwidth\" [attr.id]=\"componentId\">\n          </svg>\n      </td>\n  </tr>\n  <tr *ngIf=\"legend\">\n      <td style=\"display: block; padding: 5px;\">\n          <amexio-d3-legend *ngIf=\"legend\" [horizontal]=\"true\" (onClick)=\"onCandlestickLegendClick($event)\" [data]=\"legendArray\" style=\"height:100%;\">\n          </amexio-d3-legend>\n      </td>\n  </tr>\n</table>\n<div *ngIf=\"data.length < 1\" class=\"lmask\" style=\"padding-top:50%\">\n    </div>\n</div>",
                    styles: [""]
                },] },
    ];
    CandlestickComponent.ctorParameters = function () { return [
        { type: CommanDataService },
        { type: ChangeDetectorRef },
        { type: DeviceQueryService }
    ]; };
    CandlestickComponent.propDecorators = {
        svgwidth: [{ type: Input, args: ['width',] }],
        svgheight: [{ type: Input, args: ['height',] }],
        data: [{ type: Input }],
        tickscount: [{ type: Input, args: ['yaxis-interval',] }],
        showzeroflag: [{ type: Input, args: ['show-zero-values',] }],
        chartId: [{ type: ViewChild, args: ['chartId',] }],
        divid: [{ type: ViewChild, args: ['divid',] }],
        onLegendClick: [{ type: Output }]
    };
    return CandlestickComponent;
}(AmexioD3BaseChartComponent));
export { CandlestickComponent };
if (false) {
    /** @type {?} */
    CandlestickComponent.prototype.svgwidth;
    /** @type {?} */
    CandlestickComponent.prototype.svgheight;
    /** @type {?} */
    CandlestickComponent.prototype.data;
    /** @type {?} */
    CandlestickComponent.prototype.tickscount;
    /** @type {?} */
    CandlestickComponent.prototype.showzeroflag;
    /** @type {?} */
    CandlestickComponent.prototype.chartId;
    /** @type {?} */
    CandlestickComponent.prototype.divid;
    /** @type {?} */
    CandlestickComponent.prototype.onLegendClick;
    /** @type {?} */
    CandlestickComponent.prototype.predefinedColor;
    /** @type {?} */
    CandlestickComponent.prototype.keyArray;
    /** @type {?} */
    CandlestickComponent.prototype.transformeddata;
    /** @type {?} */
    CandlestickComponent.prototype.height;
    /** @type {?} */
    CandlestickComponent.prototype.width;
    /** @type {?} */
    CandlestickComponent.prototype.margin;
    /** @type {?} */
    CandlestickComponent.prototype.x;
    /** @type {?} */
    CandlestickComponent.prototype.y;
    /** @type {?} */
    CandlestickComponent.prototype.svg;
    /** @type {?} */
    CandlestickComponent.prototype.tooltip;
    /** @type {?} */
    CandlestickComponent.prototype.legendArray;
    /** @type {?} */
    CandlestickComponent.prototype.httpresponse;
    /** @type {?} */
    CandlestickComponent.prototype.wt;
    /**
     * @type {?}
     * @private
     */
    CandlestickComponent.prototype.myservice;
    /**
     * @type {?}
     * @private
     */
    CandlestickComponent.prototype.cdf;
    /**
     * @type {?}
     * @private
     */
    CandlestickComponent.prototype.device;
}
